File: backend/src/main/java/org/kpet/petshop/Controllers/UploadController.java
package org.kpet.petshop.Controllers;

import org.kpet.petshop.Services.FileStorageService;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MaxUploadSizeExceededException;
import org.springframework.web.multipart.MultipartFile;

import java.util.Map;

/**
 * Generic image upload used by every admin form that has an image field
 * (products, blog posts, distributors, site logo/QR/banners). Max size (15MB)
 * is enforced at the Spring multipart layer; file-type validation happens in
 * FileStorageService.
 */
@RestController
@RequestMapping("/api/upload")
public class UploadController {

    private final FileStorageService fileStorageService;

    public UploadController(FileStorageService fileStorageService) {
        this.fileStorageService = fileStorageService;
    }

    @PostMapping(consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
    public ResponseEntity<?> upload(@RequestParam("file") MultipartFile file) {
        try {
            String url = fileStorageService.store(file);
            return ResponseEntity.ok(Map.of("url", url));
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest().body(Map.of("error", e.getMessage()));
        }
    }

    @ExceptionHandler(MaxUploadSizeExceededException.class)
    public ResponseEntity<?> handleTooLarge(MaxUploadSizeExceededException e) {
        return ResponseEntity.status(HttpStatus.PAYLOAD_TOO_LARGE)
                .body(Map.of("error", "La imagen es demasiado grande. El tamaño máximo permitido es 15MB."));
    }
}

File: backend-php/app/Http/Controllers/UploadController.php
<?php

namespace App\Http\Controllers;

use App\Services\FileStorageService;
use App\Services\VideoStorageService;
use Illuminate\Http\Request;

class UploadController extends Controller
{
    private const MAX_KB = 15360; // 15MB, matches spring.servlet.multipart.max-file-size
    private const MAX_VIDEO_KB = 51200; // 50MB — short blog clips only, not full episodes

    public function upload(Request $request, FileStorageService $storage)
    {
        if (!$request->hasFile('file')) {
            return response()->json(['error' => 'No se envió ningún archivo.'], 400);
        }

        $file = $request->file('file');
        if ($file->getSize() > self::MAX_KB * 1024) {
            return response()->json(['error' => 'La imagen es demasiado grande. El tamaño máximo permitido es 15MB.'], 413);
        }

        try {
            $url = $storage->store($file);
        } catch (\InvalidArgumentException $e) {
            return response()->json(['error' => $e->getMessage()], 400);
        }

        return response()->json(['url' => $url]);
    }

    public function uploadVideo(Request $request, VideoStorageService $storage)
    {
        if (!$request->hasFile('file')) {
            return response()->json(['error' => 'No se envió ningún archivo.'], 400);
        }

        $file = $request->file('file');
        if ($file->getSize() > self::MAX_VIDEO_KB * 1024) {
            return response()->json(['error' => 'El video es demasiado grande. El tamaño máximo permitido es 50MB.'], 413);
        }

        try {
            $url = $storage->store($file);
        } catch (\InvalidArgumentException $e) {
            return response()->json(['error' => $e->getMessage()], 400);
        }

        return response()->json(['url' => $url]);
    }
}
